src/Storages/PartitionStrategy.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <Interpreters/ExpressionActions.h>
+#include <Parsers/IAST_fwd.h>
+
+#include <Processors/Chunk.h>
+
+namespace DB
+{
+
+struct PartitionStrategy
+{
+    struct PartitionExpressionActionsAndColumnName
+    {
+        ExpressionActionsPtr actions;
+        std::string column_name;
+    };
+
+    PartitionStrategy(ASTPtr partition_by_, const Block & sample_block_, ContextPtr context_);
+
+    virtual ~PartitionStrategy() = default;
+
+    virtual PartitionExpressionActionsAndColumnName getExpression() = 0;
+    virtual std::string getPath(const std::string & prefix, const std::string & partition_key) = 0;
+
+    virtual Chunk getChunkWithoutPartitionColumnsIfNeeded(const Chunk & chunk)
+    {
+        Chunk result;
+
+        for (const auto & column : chunk.getColumns())
+        {
+            result.addColumn(column);
+        }
+
+        return result;
+    }
+
+    virtual Block getBlockWithoutPartitionColumnsIfNeeded()
+    {
+        return sample_block;
+    }
+
+protected:
+    ASTPtr partition_by;
+    Block sample_block;
+    ContextPtr context;
+};
+
+struct PartitionStrategyProvider
```

Comment:
I would prefer name `PartitionStrategyFactory` here, the nuances of the name is debatable, but IMO `provider` is:

1. more vague, while this particular class just builds an instance based on input arguments.
2. most often used in CH codebase in different meaning: like a simpler interface to the underlying complex system (metrics provider, columns in dictionaries), or give access to some pre-existing pre-configured component, managed elsewhere.